src/trainers/trainer.cpp

```diff
@@ -115,6 +129,11 @@ void trainer::setup(std::unique_ptr<thread_pool> io_thread_pool) {
   // Setup I/O threads - set up before setting up the layers (input
   // layer depends on having a properly initialized thread pool)
   m_io_thread_pool = std::move(io_thread_pool);
+
+  // Set up callbacks
+  for (const auto& cb : m_callbacks) {
+    cb->setup(this);
```

Comment:
How does this compile? `callback_base::setup(trainer*)` is not a `const` member function.